autoretrieve/engine.go

```diff
@@ -0,0 +1,524 @@
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	lnk, err := e.lsys.Store(ipld.LinkContext{Ctx: ctx}, schema.Linkproto, adNode)
+	if err != nil {
+		return cid.Undef, fmt.Errorf("cannot generate advertisement link: %s", err)
+	}
+	c := lnk.(cidlink.Link).Cid
+	log := log.With("adCid", c)
+	log.Info("Stored ad in local link system")
+
+	if err := e.putLatestAdv(ctx, c.Bytes()); err != nil {
+		log.Errorw("Failed to update reference to the latest advertisement", "err", err)
+		return cid.Undef, fmt.Errorf("failed to update reference to latest advertisement: %w", err)
+	}
+	log.Info("Updated reference to the latest advertisement successfully")
+	return c, nil
+}
+
+// Publish stores the given advertisement locally via AutoretrieveEngine.PublishLocal first, then publishes
+// a message onto the gossipsub to signal the change in the latest advertisement by the provider to
+// indexer nodes.
+//
+// The publication mechanism uses legs.Publisher internally.
+// See: https://github.com/filecoin-project/go-legs
+func (e *AutoretrieveEngine) Publish(ctx context.Context, adv schema.Advertisement) (cid.Cid, error) {
+	c, err := e.PublishLocal(ctx, adv)
+	if err != nil {
+		log.Errorw("Failed to store advertisement locally", "err", err)
+		return cid.Undef, fmt.Errorf("failed to publish advertisement locally: %w", err)
+	}
+
+	// Only announce the advertisement CID if publisher is configured.
+	if e.publisher != nil {
+		log := log.With("adCid", c)
+		log.Info("Publishing advertisement in pubsub channel")
+		err = e.publisher.UpdateRoot(ctx, c)
+		if err != nil {
+			log.Errorw("Failed to announce advertisement on pubsub channel ", "err", err)
+			return cid.Undef, err
+		}
+	}
+	return c, nil
+}
+
+// PublishLatest re-publishes the latest existing advertisement to pubsub.
+func (e *AutoretrieveEngine) PublishLatest(ctx context.Context) error {
+	// Skip announcing the latest advertisement CID if there is no publisher.
+	if e.publisher == nil {
+		log.Infow("Skipped announcing the latest: remote announcements are disabled.")
+		return nil
+	}
+
+	adCid, err := e.getLatestAdCid(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get latest advertisement cid: %w", err)
+	}
+
+	if adCid == cid.Undef {
+		log.Info("Skipped announcing the latest: no previously published advertisements.")
+		return nil
+	}
+	log.Infow("Republishing latest advertisement", "cid", adCid)
+
+	return e.publisher.UpdateRoot(ctx, adCid)
+}
+
+// RegisterMultihashLister registers a provider.MultihashLister that is used to look up the
+// list of multihashes associated to a context ID. At least one such registration
+// must be registered before calls to AutoretrieveEngine.NotifyPut and AutoretrieveEngine.NotifyRemove.
+//
+// Note that successive calls to this function will replace the previous registration.
+// Only a single registration is supported.
+//
+// See: provider.Interface
+func (e *AutoretrieveEngine) RegisterMultihashLister(mhl provider.MultihashLister) {
```

Comment:
I don't think we need mutex. when initializing the engine, let the caller pass the `mhlister` handler func and set it once - it gives the added advantage of validating it exists before boot up (since the engine wont work without it) and one less complexity with mutex

Reply:
no need for this function, hence no need for locking and unlocking (blocking)